include/eosio/ship_protocol.hpp

```diff
@@ -0,0 +1,761 @@
+#pragma once
+
+#include <eosio/check.hpp>
+#include <eosio/crypto.hpp>
+#include <eosio/eosio_outcome.hpp>
+#include <eosio/fixed_bytes.hpp>
+#include <eosio/float.hpp>
+#include <eosio/name.hpp>
+#include <eosio/stream.hpp>
+#include <eosio/time.hpp>
+#include <eosio/varint.hpp>
+
+// todo: move
+namespace eosio {
+template <typename S>
+result<void> to_json(const input_stream& data, S& stream) {
+   return to_json_hex(data.pos, data.end - data.pos, stream);
+}
+} // namespace eosio
+
+namespace eosio { namespace ship_protocol {
+
+   typedef __uint128_t uint128_t;
+
+#ifdef __eosio_cdt__
+#   pragma clang diagnostic push
+#   pragma clang diagnostic ignored "-Winvalid-noreturn"
+   [[noreturn]] inline void report_error(const std::string& s) { eosio::check(false, s); }
+#   pragma clang diagnostic pop
+#else
+   [[noreturn]] inline void report_error(const std::string& s) { throw std::runtime_error(s); }
+#endif
+
+   struct extension {
+      uint16_t            type = {};
+      eosio::input_stream data = {};
+   };
+
+   EOSIO_REFLECT(extension, type, data)
+
+   enum class transaction_status : uint8_t {
+      executed  = 0, // succeed, no error handler executed
+      soft_fail = 1, // objectively failed (not executed), error handler executed
+      hard_fail = 2, // objectively failed and error handler objectively failed thus no state change
+      delayed   = 3, // transaction delayed/deferred/scheduled for future execution
+      expired   = 4, // transaction expired and storage space refunded to user
+   };
+
+   // todo: switch to eosio::result. switch to new serializer string support.
+   inline std::string to_string(transaction_status status) {
+      switch (status) {
+         case transaction_status::executed: return "executed";
+         case transaction_status::soft_fail: return "soft_fail";
+         case transaction_status::hard_fail: return "hard_fail";
+         case transaction_status::delayed: return "delayed";
+         case transaction_status::expired: return "expired";
+      }
+      report_error("unknown status: " + std::to_string((uint8_t)status));
+   }
+
+   // todo: switch to eosio::result. switch to new serializer string support.
+   inline transaction_status get_transaction_status(const std::string& s) {
+      if (s == "executed")
+         return transaction_status::executed;
+      if (s == "soft_fail")
+         return transaction_status::soft_fail;
+      if (s == "hard_fail")
+         return transaction_status::hard_fail;
+      if (s == "delayed")
+         return transaction_status::delayed;
+      if (s == "expired")
+         return transaction_status::expired;
+      report_error("unknown status: " + s);
+   }
+
+   template <typename S>
+   eosio::result<void> from_bin(transaction_status& obj, S& stream) {
+      return stream.read_raw(obj);
+   }
```

Comment:
I don't think this is needed.  This is the default behavior for scoped enums.

Reply:
removed